pkg/lambci-lambda/service.go

```diff
@@ -0,0 +1,278 @@
+		homeDir:        config.GetString("directory"),
+		runtimes:       make(map[string][]string),
+		defaultRuntime: config.GetString("default-runtime"),
+		session:        nil,
+		log:            logger,
+	}
+
+	// not setting this value can have a bad outcome for the filesystem
+	if service.homeDir == "" {
+		return nil, errors.New("configuration setting 'directory' is required")
+	}
+
+	if strings.HasPrefix(service.homeDir, "~/") {
+		usr, err := user.Current()
+		if err != nil {
+			return nil, errors.Wrap(err, "error loading current user")
+		}
+		service.homeDir = usr.HomeDir + service.homeDir[1:]
+	}
+
+	createDirIfNotExists := func(subdir string) error {
+		path := filepath.Join(service.homeDir, subdir)
+		if _, err := os.Stat(path); os.IsNotExist(err) {
+			_, err := service.Exec(fmt.Sprintf("mkdir -p %s", path))
+			if err != nil {
+				return errors.Wrap(err, "error creating lambci directories")
+			}
+		} else {
+			return errors.Wrap(err, "error checking lambci directories")
+		}
+		return nil
+	}
+
+	if err := createDirIfNotExists(taskDir); err != nil {
+		return nil, err
+	}
+	if err := createDirIfNotExists(runtimeDir); err != nil {
+		return nil, err
+	}
+	if err := createDirIfNotExists(layersDir); err != nil {
+		return nil, err
+	}
+
+	env := filepath.Join(service.homeDir, envFile)
+	if _, err := os.Stat(env); err != nil {
+		_, err = service.Exec(fmt.Sprintf("touch %s", env))
+		if err != nil {
+			return nil, errors.Wrap(err, "error creating lambci env file")
+		}
+	}
+
+	for name, config := range config.GetStringMap("runtimes") {
+
+		runtimeConfig := config.(map[string]interface{})
+		layerConfig := runtimeConfig["layers"].([]interface{})
+
+		layers := make([]string, len(layerConfig))
+		for i := 0; i < len(layerConfig); i++ {
+			layers[i] = layerConfig[i].(string)
+		}
+
+		service.runtimes[name] = layers
+	}
+
+	return service, nil
+}
+
+// Package up everything needed to install the function but don't actually
+// install it to the service. rawDir may be assumed to be a unique path for
+// this function. The package location should be determinsitically derived
+// from the rawDir path.
+// Returns: Path to the newly created package
+func (service *lambciLambda) Package(rawDir string) (string, error) {
+
+	// for lambci-lambda there is no need to create zip files
+	return rawDir, nil
+}
+
+// Install a function to the desired FaaS service. It is assumed that
+// Package() has already been called on this rawDir. The name of rawDir is
+// also the name of the function.
+func (service *lambciLambda) Install(rawDir string, env map[string]string, runtime string) error {
+
+	if runtime == "" {
+		runtime = service.defaultRuntime
+	}
+
+	// we allow no runtime for lambci as it could be provided by the container
+	if runtime != "" {
+		if _, exists := service.runtimes[runtime]; !exists {
+			return errors.Errorf("runtime '%s' does not exist in configuration", runtime)
+		}
+	}
+
+	// remove old layer
```

Comment:
This would imply that lambci cannot have more than one function installed at any given time. This might be good to document somewhere as it's different than the other providers.